### PR TITLE
WIP: GitHub download by URL

### LINKED ISSUE #638: Projects URLs should be able to view any project
Right now one can refer to a view of an already-synced project using an URL like http://app.flowhub.io/#project/rebirth/rebirth_graphs_ExtractColors - however if the project is not synced it will fail. Instead we should be able to on-demand load (a part of) a project.

This can be used for having a "Open in Flowhub" URL in documentation, sending to a friend etc.

It is desirable to be able to specify at least:

* Project/repo
* Branch/commit/tag
* Graph/component/fbpspec/filename

If feasible, we could use the same URL format as on Github. This would make it possible for people to just replace "github.com" with "flowhub.io", to open anything one looks at in Github easily.

As this will be an a feature which external consumers will refer to (in their README.md files, documentation, in emails/chat history etc), it is important that it stays stable. Hence we must have automated tests that exercise the different cases.

### PR DESCRIPTION
**Pull request by bergie**
Will example a "Open in Flowhub" mechanism where users can provide a URL to open a given GitHub repo directly in Flowhub with:

* Repo master: <https://app.flowhub.io/#github/c-base/ingress-table>
* Particular branch: <https://app.flowhub.io/#github/c-base/ingress-table/tree/foo>
* Particular file: <https://app.flowhub.io/#github/c-base/ingress-table/blob/master/graphs/FetchDat.json>

(these are essentially GitHub URLs with `https://github.com/` replaced with `https://app.flowhub.io/#github/`)

If user already has that repo downloaded, we redirect to normal project view. If not, we download and then redirect.

We will be able to use this mechanism also for the example projects, meaning that they'll get downloaded to user's IndexedDB and user can then edit them locally with persistence.

The same behavior is applied also to gist examples, which solves #229 and #475.

Fixes #638

### CODE REVIEW

File: graphs/Save.fbp [hunk 1]
```diff
@@ -42,8 +42,9 @@ SplitSpec OUT -> IN EnsureId
 StoreName OUT -> NAME Store(indexeddb/GetStore)
 WriteTransaction TRANSACTION -> TRANSACTION Store STORE -> STORE Put(indexeddb/Put)
 EnsureId OUT -> VALUE Put
-'state=ok' -> PROPERTY SetOk(objects/SetProperty)
-Put KEY -> START CreateOk(objects/CreateObject) OUT -> IN SetOk OUT -> IN MergeContext
+StoreName OUT -> TYPE CreateOk(ui/PersistedToContext)
```
Comment (jonnor) [inline graphs/Save.fbp]: What is a Persisted?

File: components/GistToProject.coffee [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    datatype: 'object'
+  c.outPorts.add 'error',
+    datatype: 'object'
+
+  noflo.helpers.WirePattern c,
+    out: ['graph', 'component', 'project']
+    async: true
+  , (data, groups, out, callback) ->
+    api = octo.api()
+    api.token data.state.user['github-token'] if data.state.user?['github-token']
```
Comment (jonnor) [inline components/GistToProject.coffee]: Don't like that a generic and should-have-small-scope component like this gets the entire state. It makes data dependencies invisible in the graph

File: components/CheckGistExists.coffee [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  c.outPorts.add 'existing',
+    datatype: 'array'
+  c.outPorts.add 'new',
+    datatype: 'object'
+
+  noflo.helpers.WirePattern c,
+    out: ['existing', 'new']
+    async: true
+  , (data, groups, out, callback) ->
+    unless data.state?.projects?.length
```
Comment (jonnor) [inline components/CheckGistExists.coffee]: Same, I think this component should just get `projects`

File: graphs/Save.fbp (hunk 1 above)
Comment (bergie) [inline graphs/Save.fbp]: "Persisted" is whatever we just saved ... graph, runtime, component, project, spec.

File: components/GistToProject.coffee (hunk 2 above)
Comment (bergie) [inline components/GistToProject.coffee]: Fair point, we can plug a `GetActionValues` in front of this so we only pick the pieces we need.

**Comment by bergie** [comment]
Here is how the high-level architecture should look like. With a bit of subgraphing we can make `ui/main` look like this:

screenshot 2017-02-23 at 17 06 39

**Pull request closed without merging**